### PR TITLE
SP-8750: whitelist lightness and sensor statueses

### PR DESCRIPTION
**Pull request by przemyslawfierek-silvair**

### CODE REVIEW

**Comment by michallowasrzechonek-silvair** [comment]
Meh. I'd rather make it configurable via Amqp1 API.

File: mesh/model.c [hunk 1]
```diff
@@ -884,14 +884,44 @@ static void send_dev_key_msg_rcvd(struct mesh_node *node, uint8_t ele_idx,
 					   size, data);
 }
 
+static inline bool match_opcode(const void *a, const void *b)
+{
+	return *(const uint32_t *)a == *(const uint32_t *)b;
```
Comment (michallowasrzechonek-silvair) [inline mesh/model.c]: `L_PTR_TO_UINT`

Comment (przemyslawfierek-silvair) [reply mesh/model.c]: This is correct. `L_PTR_TO_UINT` is to extract `unsigned int` from pointer value.

```#define L_PTR_TO_UINT(p) ((unsigned int) ((uintptr_t) (p)))```

File: mesh/amqp.c [hunk 2]
```diff
@@ -1172,3 +1175,21 @@ bool mesh_amqp_is_ready(struct mesh_amqp *amqp)
 {
 	return amqp->amqp_state == MESH_AMQP_STATE_CONNECTED;
 }
+
+bool mesh_amqp_opcodes_whitelist_add(struct mesh_amqp *amqp, uint32_t opcode)
+{
+	uint32_t *_opcode = l_new(uint32_t, 1);
+	*_opcode = opcode;
+
+	return l_queue_push_tail(amqp->opcodes_whitelist, _opcode);
```
Comment (michallowasrzechonek-silvair) [inline mesh/amqp.c]: I think you could avoid `l_new` by storing opcodes directly, via `L_UINT_TO_PTR`/`L_PTR_TO_UINT`, see model.c mod->subs for an example.

File: mesh/model.c (hunk 1 above)
Comment (michallowasrzechonek-silvair) [reply mesh/model.c]: Right, but see the other comment. Sorry I wasn't clear.

File: mesh/node.c [hunk 3]
```diff
@@ -2754,6 +2781,9 @@ static void setup_node_interface(struct l_dbus_interface *iface)
 
 static void setup_amqp_interface(struct l_dbus_interface *iface)
 {
+	l_dbus_interface_method(iface, "WhitelistAddOpcode", 0,
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Maybe expose this as a writable property (list of numbers)?

File: mesh/model.c [hunk 4]
```diff
@@ -884,14 +884,47 @@ static void send_dev_key_msg_rcvd(struct mesh_node *node, uint8_t ele_idx,
 					   size, data);
 }
 
+static inline bool match_opcode(const void *a, const void *b)
+{
+	return L_PTR_TO_UINT(a) == L_PTR_TO_UINT(b);
```
Comment (michallowasrzechonek-silvair) [inline mesh/model.c]: I think you should be able to compare directly, see `simple_match`.

File: mesh/amqp.c [hunk 5]
```diff
@@ -1172,3 +1175,8 @@ bool mesh_amqp_is_ready(struct mesh_amqp *amqp)
 {
 	return amqp->amqp_state == MESH_AMQP_STATE_CONNECTED;
 }
+
+struct l_queue *mesh_amqp_get_opcodes_whitelist(struct mesh_amqp *amqp)
```
Comment (michallowasrzechonek-silvair) [inline mesh/amqp.c:1179]: I'd remove this function, see below.

File: mesh/model.c [hunk 6]
```diff
@@ -884,14 +884,47 @@ static void send_dev_key_msg_rcvd(struct mesh_node *node, uint8_t ele_idx,
...
 static void send_fd_msg_rcvd(uint8_t ele_idx,
 				 uint16_t src, uint16_t dst,
 				 const struct mesh_virtual *virt,
 				 uint16_t app_idx,
 				 uint16_t size, const uint8_t *data,
 				 send_callback_t send_callback, void *user_data)
 {
-	struct fd_msg *msg = fd_msg_new(data, size);
+	struct fd_msg *msg;
+
```
Comment (michallowasrzechonek-silvair) [inline mesh/model.c:914]: `struct mesh_amqp *amqp = user_data;`

File: mesh/model.c [hunk 7]
```diff
@@ -884,14 +884,47 @@ static void send_dev_key_msg_rcvd(struct mesh_node *node, uint8_t ele_idx,
...
+	if (send_callback == send_amqp) {
+		struct l_queue *whitelist =
+				mesh_amqp_get_opcodes_whitelist(user_data);
+
+		if (!whitelist)
+			return;
+
+		if (IS_GROUP(dst) &&
+				!is_opcode_whitelisted(whitelist, data, size))
```
Comment (michallowasrzechonek-silvair) [inline mesh/model.c]: How about:

```C
struct mesh_amqp *amqp = user_data;

if (send_callback == send_amqp) {
    uint32_t opcode;
    uint16_t opcode_len;

    if (!mesh_model_opcode_get(buf, size, &opcode, &opcode_len)) {
        l_warn("Failed to get an opcode. Buffer too short");
        return;
   }

    if (IS_GROUP(dst) && !l_queue_find(amqp->opcodes_whitelist, simple_match, L_UINT_TO_PTR(opcode)))
        return;
    ...
}
```

File: mesh/node.c [hunk 8]
```diff
@@ -2692,6 +2691,77 @@ static struct l_dbus_message *amqp_identity_setter(struct l_dbus *dbus,
...
+						"Internal error - AMQP");
+
+	whitelist = mesh_amqp_get_opcodes_whitelist(node->amqp);
+
+	if (!whitelist)
+		return dbus_error(msg, MESH_ERROR_DOES_NOT_EXIST,
+					"Internal error - Opcodes Whitelist");
+
+	if (strcmp(sender, node->owner))
+		return dbus_error(msg, MESH_ERROR_NOT_AUTHORIZED, NULL);
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c:2720]: 👍🏻

File: mesh/node.c [hunk 9]
```diff
@@ -2692,6 +2691,77 @@ static struct l_dbus_message *amqp_identity_setter(struct l_dbus *dbus,
...
+
+	if (!l_dbus_message_iter_get_variant(value, "au", &iter))
+		return dbus_error(msg, MESH_ERROR_INVALID_ARGS,
+						"List of opcodes expected");
+
+	l_queue_clear(whitelist, NULL);
+
+	while (l_dbus_message_iter_next_entry(&iter, &opcode)) {
+		l_info("Whitelisted opcode: 0x%X", opcode);
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c:2734]: Maybe verify that incoming number is a legitimate opcode?

File: mesh/node.c [hunk 10]
```diff
@@ -2692,6 +2691,88 @@ static struct l_dbus_message *amqp_identity_setter(struct l_dbus *dbus,
 	return NULL;
 }
 
+static bool is_opcode_valid(uint32_t opcode)
+{
+	if (opcode > 0xFFFFFFul || (opcode >> 16) == 0x7F)
```
Comment (michallowasrzechonek-silvair) [inline mesh/node.c]: Why not use `mesh_model_opcode_get`?

File: mesh/amqp.c (hunk 5 above)
Comment (michallowasrzechonek-silvair) [inline mesh/amqp.c:1179]: You don't need that function now ;)

**Review by michallowasrzechonek-silvair — APPROVED** [review]
Provisional LGTM, just do something about `is_opcode_valid`.

👍🏻

File: mesh/node.c (hunk 10 above)
Comment (przemyslawfierek-silvair) [reply mesh/node.c]: You mean `mesh_model_opcode_set`, right?

**Review by michallowasrzechonek-silvair — APPROVED** [review]

**Pull request merged**